packages/outline-view/src/browser/outline-view-tree.ts

```diff
@@ -0,0 +1,52 @@
+/*
+* Copyright (C) 2017 TypeFox and others.
+*
+* Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+*/
+
+import { injectable, inject } from "inversify";
+import { Tree, ICompositeTreeNode, ITreeNode } from "@theia/core/lib/browser";
+import { OutlineSymbolInformationNode, OutlineViewManager } from './outline-view-manager';
+
+@injectable()
+export class OutlineViewTree extends Tree {
+
+    constructor( @inject(OutlineViewManager) protected readonly outlineViewManager: OutlineViewManager) {
+        super();
+
+        outlineViewManager.onDidChangeOutline(nodes => {
+            this.setOutlineTree(nodes);
+        });
+
+        this.root = <ICompositeTreeNode>{
+            visible: false,
+            id: 'theia-outline-view-widget',
+            name: 'OutlineView',
+            children: [],
+            parent: undefined
+        };
+    }
+
+    protected setOutlineTree(tree: OutlineSymbolInformationNode[]) {
+        // update expansionstate and selectionstate
+        const nodes = this.setExpansionState(tree);
+        this.root = nodes[0];
+        this.refresh();
+    }
+
+    protected setExpansionState(nodes: ITreeNode[]): ITreeNode[] {
```

Comment:
I think the tree should be changed to spread new nodes in current nodes to preserve UI state then clients don't need to care. We do it like that for all trees, except here. The problem will arise when there is new UI state which is not preserved here. It could be done separately.

Reply:
Isn't the whole point of using virtual doms, to not do such complex mutation stuff?

Reply:
Maybe I'm missing something. I don't know how the virtual dom can preserve the UI state of nodes, as node expansion/selection, between rerendering. Even if we use virtual nodes directly as the tree state, instead of JSON-objects as now, the issue remains: at some point, we have to decide whether we have to create a new virtual node or spread data in the existing.